### PR TITLE
cylc unregister: skip suites with port files

### LINKED ISSUE #1796: cylc unregister: running suites can be unregistered
It is currently possible to unregister a running suite with `cylc unregister`, the suite remains visible to `cylc scan` but `cylc shutdown` yields `[client-connect] DENIED (privilege 'state-totals' < 'shutdown')`.

### PR DESCRIPTION
**Pull request by matthewrmshin**
Also ensure that `cylc unregister bar` does not unregister a suite
called `foobar`.

Close #1796.

### CODE REVIEW

**Comment by matthewrmshin** [comment]
@hjoliver @arjclark please review.

**Comment by hjoliver** [comment]
Branch needs updating.

File: bin/cylc-unregister [hunk 1]
```diff
@@ -56,55 +58,43 @@ def main():
...
+            continue
+        # recursively remove empty superdirs
+        tmp = path
+        while True:
+            tmp = os.path.split(tmp)[0]
+            try:
+                os.rmdir(tmp)
+            except OSError:
+                break
             else:
```
Comment (hjoliver) [inline bin/cylc-unregister]: This block (in master too) is a bit too simplistic to do what it's supposed to - i.e. remove all empty directories left by unregistering and deleting a bunch of suites.  To see this, try `cylc import examples $TMPDIR` then `cylc unreg -d cylc-.*` and look in `$TMPDIR`.  However, I doubt that anyone stores suites in directory trees that reflect the suite name hierarchy as I originally intended (e.g. `nwp.test1` and `nwp.test2` stored in `nwp/test1/` and `nwp/test2`) - certainly Rose users don't - and unregisters them en-masse ... so maybe we should just remove this feature.

File: bin/cylc-unregister [hunk 2]
```diff
@@ -56,55 +58,43 @@ def main():
...
+                print 'Removed empty directory:', tmp
+    sys.exit(bool(skipped_list))
 
 if __name__ == "__main__":
     try:
         main()
     except Exception as exc:
         if cylc.flags.debug:
-            raise
+            traceback.print_exc()
```
Comment (hjoliver) [inline bin/cylc-unregister]: Should we do this in all commands?

File: lib/cylc/registration.py [hunk 3]
```diff
@@ -150,26 +152,30 @@ def get_list(self, regfilter=None):
 
     def unregister(self, exp):
         """Un-register a suite."""
-        suitedirs = []
-        for key in self.list_all_suites():
-            if re.search(exp + '$', key):
+        unregistered_list = set()
+        skipped_list = set()
```
Comment (hjoliver) [inline lib/cylc/registration.py]: `unregistered_set` etc.?

File: tests/registration/03-skip-run.t [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+[cylc]
+    [[event hooks]]
+        timeout = PT1M
+        abort on timeout = True
+[scheduling]
+    [[dependencies]]
+        graph = foo
+[runtime]
+    [[foo]]
+        script = printenv FOO
```
Comment (hjoliver) [inline tests/registration/03-skip-run.t]: Why `printenv FOO`?

Comment (hjoliver) [inline tests/registration/03-skip-run.t]: Ok - to make the task fail, but `/bin/false` would be more obvious (no need to check if `$FOO` is defined anywhere).

File: bin/cylc-unregister (hunk 2 above)
Comment (matthewrmshin) [inline bin/cylc-unregister]: I was doing this here to allow me to print out a better trace back, and forgot to change it back. It should probably be done for all commands in a separate change.

File: tests/registration/03-skip-run.t (hunk 4 above)
Comment (matthewrmshin) [inline tests/registration/03-skip-run.t]: Sorry, leaking one of my toy suites here. (The purpose of the toy suite was to test broadcast of an environment variable after a failure.)

**Comment by matthewrmshin** [comment]
All comments addressed. Branch re-based.

**Comment by hjoliver** [comment]
Review 1 - good, test battery passing.

**Comment by arjclark** [comment]
Looks OK to me. No problems from the test-battery in my environment.

**Pull request merged**